### PR TITLE
Move terms in SpeciesAtom to references

### PR DESCRIPTION
**Pull request by rprospero**
The pointers in SpeciesAtom have been converted into references.  References were chosen over safe pointers because the actual values are contained in a `std::vector` in `Species`, so the lifetime of the vector and of the pointer would conflict.

### CODE REVIEW

**Comment by rprospero** [comment]
As a preemptive comment, many places in here have explicit type declarations, instead of `auto`.  This is necessary since the vectors contain reference wrappers and declaring a reference to the exact type is necessary to unwrap it.

File: src/classes/species.cpp [hunk 1]
```diff
@@ -107,9 +107,9 @@ bool Species::checkSetUp()
         }
 
         // Check each Bond for two-way consistency
-        for (const auto *bond : i->bonds())
+        for (const SpeciesBond &bond : i->bonds())
         {
-            SpeciesAtom *partner = bond->partner(i);
+            SpeciesAtom *partner = bond.partner(i);
```
Comment (trisyoungs) [inline src/classes/species.cpp:112]:
```suggestion
            auto *partner = bond.partner(i);
```

File: src/classes/species_intra.cpp [hunk 2]
```diff
@@ -240,7 +240,13 @@ SpeciesAngle &Species::addAngle(SpeciesAtom *i, SpeciesAtom *j, SpeciesAtom *k)
     }
 
     // OK to add new angle
-    angles_.emplace_back(i, j, k).setParent(this);
+
+    // We can't use emplace_back since SpeciesAngle needs
+    // to derference its `this` pointer in the constructor
```
Comment (trisyoungs) [inline src/classes/species_intra.cpp:245]:
```suggestion
    // to dereference its `this` pointer in the constructor
```

File: src/classes/species_intra.cpp [hunk 3]
```diff
@@ -240,7 +240,13 @@ SpeciesAngle &Species::addAngle(SpeciesAtom *i, SpeciesAtom *j, SpeciesAtom *k)
...
+    // to update the SpeciesAtoms.
+    SpeciesAngle tmp = SpeciesAngle(i, j, k);
+    angles_.push_back(std::move(tmp));
```
Comment (trisyoungs) [inline src/classes/species_intra.cpp:248]: My lack of knowledge on move semantics may fail me here, but can this be one-lined (i.e. `angles_.push_back(std::move(SpeciesAngle(i,j,k)))` or does `std::move` dislike objects created in-line?

File: src/classes/species_intra.cpp [hunk 4]
```diff
@@ -465,11 +471,11 @@ void Species::generateAttachedAtomLists()
     for (auto &torsion : torsions_)
     {
         // Grab relevant Bond (if it exists)
-        SpeciesBond *jk = torsion.j()->hasBond(torsion.k());
+        OptionalReferenceWrapper<SpeciesBond> jk = torsion.j()->hasBond(torsion.k());
```
Comment (trisyoungs) [inline src/classes/species_intra.cpp:474]:
```suggestion
        auto jk = torsion.j()->hasBond(torsion.k());
```

File: src/data/ff.cpp [hunk 5]
```diff
@@ -477,7 +477,7 @@ bool Forcefield::assignIntramolecular(Species *sp, int flags) const
             for (int indexJ = 0; indexJ < i->nBonds() - 2; ++indexJ)
             {
                 // Get SpeciesAtom 'j'
-                SpeciesAtom *j = i->bond(indexJ)->partner(i);
+                SpeciesAtom *j = i->bond(indexJ).partner(i);
```
Comment (trisyoungs) [inline src/data/ff.cpp:480]:
```suggestion
                auto *j = i->bond(indexJ).partner(i);
```

File: src/data/ff.cpp [hunk 6]
```diff
@@ -488,7 +488,7 @@ bool Forcefield::assignIntramolecular(Species *sp, int flags) const
                 for (int indexK = indexJ + 1; indexK < i->nBonds() - 1; ++indexK)
                 {
                     // Get SpeciesAtom 'k'
-                    SpeciesAtom *k = i->bond(indexK)->partner(i);
+                    SpeciesAtom *k = i->bond(indexK).partner(i);
```
Comment (trisyoungs) [inline src/data/ff.cpp:491]:
```suggestion
                    auto *k = i->bond(indexK).partner(i);
```

File: src/data/ff.cpp [hunk 7]
```diff
@@ -499,7 +499,7 @@ bool Forcefield::assignIntramolecular(Species *sp, int flags) const
                     for (int indexL = indexK + 1; indexL < i->nBonds(); ++indexL)
                     {
                         // Get SpeciesAtom 'l'
-                        SpeciesAtom *l = i->bond(indexL)->partner(i);
+                        SpeciesAtom *l = i->bond(indexL).partner(i);
```
Comment (trisyoungs) [inline src/data/ff.cpp:502]:
```suggestion
                        auto *l = i->bond(indexL).partner(i);
```

File: src/neta/connection.cpp [hunk 8]
```diff
@@ -218,10 +218,8 @@ int NETAConnectionNode::score(const SpeciesAtom *i, RefList<const SpeciesAtom> &
         if (nHydrogensValue_ >= 0)
         {
             // Count number of hydrogens attached to this atom
-            auto nH = 0;
-            for (const auto *bond : j->bonds())
-                if (bond->partner(j)->element()->Z() == ELEMENT_H)
-                    ++nH;
+            auto nH = std::count_if(j->bonds().begin(), j->bonds().end(),
+                                    [j](const SpeciesBond &bond) { return bond.partner(j)->element()->Z() == ELEMENT_H; });
```
Comment (trisyoungs) [inline src/neta/connection.cpp:222]: Nice.

File: src/neta/presence.cpp [hunk 9]
```diff
@@ -160,10 +160,8 @@ int NETAPresenceNode::score(const SpeciesAtom *i, RefList<const SpeciesAtom> &av
         if (nHydrogensValue_ >= 0)
         {
             // Count number of hydrogens attached to this atom
-            auto nH = 0;
-            for (const auto *bond : j->bonds())
-                if (bond->partner(j)->element()->Z() == ELEMENT_H)
-                    ++nH;
+            auto nH = std::count_if(j->bonds().begin(), j->bonds().end(),
+                                    [j](const SpeciesBond &bond) { return bond.partner(j)->element()->Z() == ELEMENT_H; });
```
Comment (trisyoungs) [inline src/neta/presence.cpp:164]: ++nice

**Review by trisyoungs — CHANGES_REQUESTED** [review]
This all looks pretty good - makes a lot of stuff look a lot neater.  I've made a handful of suggestions, but the vast majority are me adding uses of `auto` that I missed in my original sweep.

File: src/classes/species_intra.cpp (hunk 3 above)
Comment (rprospero) [reply src/classes/species_intra.cpp:248]: That worked just fine.

**Review by trisyoungs — APPROVED** [review]
This can be merged whenever you're ready.  The pipeline failure is to do with the website, and which can be ignored.

**Pull request merged**